lib/services/rally/rally_hierarchical_requirement_resource.rb

```diff
@@ -200,6 +202,62 @@ def map_requirement(parent_id, release_id, aha_requirement)
     attributes
   end
 
+  def maybe_add_tags_to_object(attributes, aha_object)
+    object_tags = aha_object.tags
+
+    if object_tags && !object_tags.empty?
+      attributes[:Tags] = get_or_create_tag_references(object_tags)
+    end
+  rescue AhaService::RemoteError => e
+    logger.error("Failed to create tag #{tag_name}: #{e.message}")
+  end
+
+  def get_or_create_tag_references(tags)
+    # Rally doesn't want escaped quotes or parens
+    query_params = build_tag_query(tags).gsub(" ", "%20").gsub("=", "%3D")
+    url = rally_secure_url_without_workspace("/tag?query=#{query_params}")
+    process_response http_get(url) do |document|
+      results = document.QueryResult.Results
+      tag_refs = []
+      results.each { |result| tag_refs << {_ref: result._ref} }
+      results_names = results.map{|obj| obj._refObjectName}
+      tags_to_create = tags.select{|tag| !results_names.include? tag}
+      created_tags = create_tags tags_to_create
+
+      return tag_refs + created_tags
+    end
+  end
+
+  # The rally api requires that queries have paren nesting. EX:
+  # (( (Name="something") OR (name="something else") ) OR (name="blah"))
+  def build_tag_query(tags, i=0, query="")
+    return query if tags.empty? || i == tags.length
+    # The query must contain spaces surrounding the `=`
+    current_query = "(Name = \"#{tags[i]}\")"
+    if i == 0
+      return build_tag_query(tags, i+1, current_query)
+    else
+      return build_tag_query(tags, i+1, "(#{query} OR #{current_query})")
```

Comment:
This would be cleaner with inject:

`tags.map{|tag| "(Name = \"#{tag}\")" }.inject {|current, n| "(#{current} OR #{n})"}`